frameworks/modular/module.h

```diff
@@ -0,0 +1,296 @@
+
+/* Subsystem API prototype name */
+#define odp_api_proto(subsystem, api) __odp_api(subsystem, api ## _proto_t)
+
+/* Subsystem API declaration */
+#define ODP_SUBSYSTEM_API(name, _return, api, ...)		\
+	extern _return __odp_api(name, api)(__VA_ARGS__);	\
+	typedef _return (*odp_api_proto(name, api))(__VA_ARGS__)
+
+/* Subsystem API stubs are weak */
+#define ODP_SUBSYSTEM_API_STUB(name, api)			\
+	__attribute__((weak)) __odp_api(name, api)
+
+/* In case a subsystem module are built as static libraries(.a)
+ * or preload dynamic libraries(.so), the module can use this
+ * macro to override the APIs weak stubs.
+ */
+#define ODP_SUBSYSTEM_API_OVERRIDE(name, api, _alias)		\
+	__attribute__((alias(#_alias))) __odp_api(name, api)
+
+#define odp_subsystem_constructor(name)				\
+	do {							\
+		odp_rwlock_init(&__subsystem(name).lock);	\
+		list_head_init(&__subsystem(name).modules);	\
+		__subsystem(name).active = NULL;		\
+	} while (0)
+
+#define ODP_SUBSYSTEM_CONSTRUCTOR(name)				\
+	static void __attribute__((constructor(101)))		\
+		odp_ ## name ## _subsystem_constructor(void)
+
+#define odp_subsystem_lock(access, name)			\
+	odp_rwlock_ ## access ## _lock(&__subsystem(name).lock)
+
+#define odp_subsystem_unlock(access, name)			\
+	odp_rwlock_ ## access ## _unlock(&__subsystem(name).lock)
+
+/* Base class to all inherited subsystem module classes */
+struct odp_module_base {
+	struct list_node list;
+	const char *name;
+	void *handler; /* DSO */
+	int (*init_local)(void);
+	int (*term_local)(void);
+	int (*init_global)(void);
+	int (*term_global)(void);
+};
+
+/* It is required to define subsystem module class with the
+ * base class as its 1st member and named as "base", and also
+ * use ODP_MODULE_CLASS(subsystem) to create the association
+ * between module class name and subsystem name, like:
+ *
+ * typedef ODP_MODULE_CLASS(subsystem) {
+ *	odp_module_base_t base;
+ *	...new members...
+ * } new_module_t; // Here pick the name you like freely
+ *
+ * It also supports forward declaration like:
+ *
+ * // Forward declaration
+ * typedef ODP_MODULE_CLASS(subsystem) new_module_t;
+ * // Laterly comes the definition
+ * ODP_MODULE_CLASS(subsystem) {
+ *	odp_module_base_t base;
+ *	...new members...
+ * }
+ *
+ * Then in preprocessor macros when we have the subsystem name
+ * we can recover the module class type information, like:
+ *
+ * #define MACRO(subsystem)
+ * do {
+ *	ODP_MODULE_CLASS(subsystem) *mod = NULL;
+ *	odp_subsystem_foreach_module(subsystem, mod) {
+ *		mod->xxx; // access the module class
+ *	}
+ * } while(0)
+ */
+#define ODP_MODULE_CLASS(subsystem) struct odp_ ## subsystem ## _module
+
+/* Below macros assume that all subsystem module classes have
+ * odp_module_base_t as their 1st member named "base".
+ *
+ * This greatly reduces the complexity for module list iteration
+ * and module pointer recovery from its list_node member by a forced
+ * type conversion instead of complex calls to container_of() etc.
```

Comment:
No, required programmer to write derived data struct like:

```C
ODP_MODULE_CLASS(subsystem) {
         odp_module_base_t base;
         ...new members...
}
```
But former macro technique can guarantee this, if we can tolerant mismatch braces, and better thing is that all derived data structure can access members in base without ".base.*" accesser.

```C
#define MODULE_CLASS(subsystem)                                \
       struct subsystem ## _module {                           \
               struct list_node list;                          \
               const char *name;                               \
               void *handler; /* DSO */                        \
               int (*init_local)(void);                        \
               int (*term_local)(void);                        \
               int (*init_global)(void);                       \
               int (*term_global)(void);                       \
```
And programmer can define their data structure like:
```C
typedef MODULE_CLASS(subsystem)
        ...new members...
        ...new members...
} new_module_t;
```